app/src/test/java/net/aiscope/gdd_app/repository/SampleRepositorySharedPreferenceTest.kt

```diff
@@ -143,18 +145,47 @@ class SampleRepositorySharedPreferenceTest {
         val yesterday = Calendar.getInstance()
         yesterday.add(Calendar.DAY_OF_YEAR, -1)
 
-        val todaySample = Sample(todayId, HOSPITAL_ID, MICROSCOPIST, status = SampleStatus.ReadyToUpload, createdOn = today)
+        val todaySample = Sample(todayId, HOSPITAL_ID, MICROSCOPIST, DISEASE, status = SampleStatus.ReadyToUpload, createdOn = today)
         val todaySampleJson = """{"id":"1112","healthFacility":"H_St_Pau","status":"ReadyToUpload","createdOn":"$today"}"""
-        val yesterdaySample = Sample(yesterdayId, HOSPITAL_ID, MICROSCOPIST, status = SampleStatus.ReadyToUpload, createdOn = yesterday)
+        val yesterdaySample = Sample(yesterdayId, HOSPITAL_ID, MICROSCOPIST, DISEASE, status = SampleStatus.ReadyToUpload, createdOn = yesterday)
         val yesterdaySampleJson = """{"id":"1113","healthFacility":"H_St_Pau","status":"ReadyToUpload","createdOn":"$yesterday"}"""
 
         whenever(gson.fromJson<SampleDto>(eq(todaySampleJson), any<Type>())).thenReturn(todaySample.toDto())
         whenever(gson.fromJson<SampleDto>(eq(yesterdaySampleJson), any<Type>())).thenReturn(yesterdaySample.toDto())
 
         whenever(store.all()).thenReturn(listOf(sampleOnlyRequiredJson, todaySampleJson, yesterdaySampleJson))
 
-        val sample = subject.last()
+        val sample = subject.lastSaved()
 
         assert(sample == todaySample)
     }
+
+    @Test
+    fun `should give last unfinished sample`() = coroutinesTestRule.runBlockingTest {
+        val todayId = "1112"
+        val yesterdayId = "1113"
+        val twoDaysId = "1114"
+        val today = Calendar.getInstance()
+        val yesterday = Calendar.getInstance()
+        yesterday.add(Calendar.DAY_OF_YEAR, -1)
+        val twoDays = Calendar.getInstance()
+        twoDays.add(Calendar.DAY_OF_YEAR, -2)
+
+        val todaySample = Sample(todayId, HOSPITAL_ID, MICROSCOPIST, DISEASE, status = SampleStatus.ReadyToUpload, createdOn = today)
+        val todaySampleJson = """{"id":"1112","healthFacility":"H_St_Pau","status":"ReadyToUpload","createdOn":"$today"}"""
+        val yesterdaySample = Sample(yesterdayId, HOSPITAL_ID, MICROSCOPIST, DISEASE, status = SampleStatus.Incomplete, createdOn = yesterday)
+        val yesterdaySampleJson = """{"id":"1113","healthFacility":"H_St_Pau","status":"Incomplete","createdOn":"$yesterday"}"""
+        val twoDaysSample = Sample(twoDaysId, HOSPITAL_ID, MICROSCOPIST, DISEASE, status = SampleStatus.Incomplete, createdOn = twoDays)
+        val twoDaysSampleJson = """{"id":"1114","healthFacility":"H_St_Pau","status":"Incomplete","createdOn":"$twoDays"}"""
+
+        whenever(gson.fromJson<SampleDto>(eq(todaySampleJson), any<Type>())).thenReturn(todaySample.toDto())
+        whenever(gson.fromJson<SampleDto>(eq(yesterdaySampleJson), any<Type>())).thenReturn(yesterdaySample.toDto())
+        whenever(gson.fromJson<SampleDto>(eq(twoDaysSampleJson), any<Type>())).thenReturn(twoDaysSample.toDto())
```

Comment:
I feel like all this block could be made shorter and easier to understand by extracting the setup in a method, something like this:

```kotlin
    private fun givenSampleAndJson(id: String, date: Calendar, status: SampleStatus): Pair<Sample, String> {
        val sample = Sample(id, HOSPITAL_ID, MICROSCOPIST, DISEASE, status = SampleStatus.ReadyToUpload, createdOn = date)
        val sampleJson = """{"id":"$id","healthFacility":"H_St_Pau","status":"$status","createdOn":"$date"}"""
        whenever(gson.fromJson<SampleDto>(eq(sampleJson), any<Type>())).thenReturn(sample.toDto())

        return sample to sampleJson
    }
```

Then, we could have only this: 

```kotlin
 		val today = Calendar.getInstance()
        val yesterday = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, -1) }
        val twoDays = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, -2) }
        
        val (_, todaySampleJson) = givenSampleAndJson("1112", today, SampleStatus.ReadyToUpload)
        val (yesterdaySample, yesterdaySampleJson) = givenSampleAndJson("1113", yesterday, SampleStatus.Incomplete)
        val (_, twoDaysSampleJson) = givenSampleAndJson("1114", twoDays, SampleStatus.Incomplete)
```

What do you think?

Reply:
Ah yeah good idea! Much more concise. 
I did rename the method to `mockSampleAndJson` to make it more clear that there is also mocking going on in this function, becomes a bit magic otherwise